### PR TITLE
[MI-2010]: API to create subscriptions

### PR DESCRIPTION
**Pull request by ayusht2810**

### CODE REVIEW

File: server/constants/constants.go [hunk 1]
```diff
@@ -29,6 +29,14 @@ const (
 	Workitems          = "_workitems"
 	Edit               = "edit"
 
+	// Subscription constants
+	PublisherID      = "tfs"
+	ConsumerId       = "webHooks"
```
Comment (Kshitij-Katiyar) [inline server/constants/constants.go:34]: This constants are not making sense consumerID is equal to webhooks and consumerActionID is http request ?
```suggestion
	ConsumerID       = "webHooks"
```

File: server/constants/constants.go [hunk 2]
```diff
@@ -29,6 +29,14 @@ const (
...
+	ConsumerActionId = "httpRequest"
```
Comment (Kshitij-Katiyar) [inline server/constants/constants.go:35]:
```suggestion
	ConsumerActionID = "httpRequest"
```

File: server/constants/messages.go [hunk 3]
```diff
@@ -36,4 +38,9 @@ const (
 	InvalidAuthState                = "Invalid oauth state, please try again"
 	UnableToCheckIfAlreadyConnected = "Unable to check if user account is already connected"
 	AuthAttemptExpired              = "Authentication attempt expired, please try again"
+	ErrorFetchSubscriptionList      = "Error in fetching subscription list"
```
Comment (Kshitij-Katiyar) [inline server/constants/messages.go:41]:
```suggestion
	FetchSubscriptionListError      = "Error in fetching subscription list"
```

File: server/constants/messages.go [hunk 4]
```diff
@@ -36,4 +38,9 @@ const (
...
+	ErrorCreateSubscription         = "Error in creating subscription"
```
Comment (Kshitij-Katiyar) [inline server/constants/messages.go:42]:
```suggestion
	CreateSubscriptionError         = "Error in creating subscription"
```

File: server/constants/routes.go [hunk 5]
```diff
@@ -2,18 +2,21 @@ package constants
...
+	APIPrefix                    = "/api/v1"
+	WildRoute                    = "{anything:.*}"
+	PathOAuthConnect             = "/oauth/connect"
+	PathOAuthCallback            = "/oauth/complete"
+	PathGetAllLinkedProjects     = "/project/link"
+	PathUnlinkProject            = "/project/unlink"
+	PathUser                     = "/user"
+	PathCreateTasks              = "/tasks"
+	PathLinkProject              = "/link"
+	PathCreateSubscriptions      = "/subscriptions"
```
Comment (Kshitij-Katiyar) [inline server/constants/routes.go]: No need to have 2 constant with same values 
```suggestion
	PathSubscriptions      = "/subscriptions"
```

File: server/plugin/api.go [hunk 6]
```diff
@@ -112,7 +113,8 @@ func (p *Plugin) handleLink(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	if p.IsProjectLinked(projectList, serializers.ProjectDetails{OrganizationName: body.Organization, ProjectName: body.Project}) {
+	_, isProjectLinked := p.IsProjectLinked(projectList, serializers.ProjectDetails{OrganizationName: body.Organization, ProjectName: body.Project})
```
Comment (Kshitij-Katiyar) [inline server/plugin/api.go:116]:
```suggestion
	if _, isProjectLinked := p.IsProjectLinked(projectList, serializers.ProjectDetails{OrganizationName: body.Organization, ProjectName: body.Project}); isProjectLinked {
```

File: server/plugin/api.go [hunk 7]
```diff
@@ -284,6 +287,83 @@ func (p *Plugin) handleCreateTask(w http.ResponseWriter, r *http.Request) {
...
+	}
+
+	subscriptionList, err := p.Store.GetAllSubscriptions(mattermostUserID)
+	if err != nil {
+		p.API.LogError(constants.ErrorFetchSubscriptionList, "Error", err.Error())
+		p.handleError(w, r, &serializers.Error{Code: http.StatusInternalServerError, Message: err.Error()})
+		return
+	}
+
+	_, isSubscriptionPresent := p.IsSubscriptionPresent(subscriptionList, serializers.SubscriptionDetails{OrganizationName: body.Organization, ProjectName: body.Project, ChannelID: channel.Id, EventType: body.EventType})
```
Comment (Kshitij-Katiyar) [inline server/plugin/api.go:333]:
```suggestion
	if _, isSubscriptionPresent := p.IsSubscriptionPresent(subscriptionList, serializers.SubscriptionDetails{OrganizationName: body.Organization, ProjectName: body.Project, ChannelID: channel.Id, EventType: body.EventType}); isSubscriptionPresent {
```

File: server/plugin/api.go [hunk 8]
```diff
@@ -284,6 +287,83 @@ func (p *Plugin) handleCreateTask(w http.ResponseWriter, r *http.Request) {
...
+	if isSubscriptionPresent {
+		p.API.LogError(constants.SubscriptionAlreadyPresent, "Error")
+		p.handleError(w, r, &serializers.Error{Code: http.StatusBadRequest, Message: constants.SubscriptionAlreadyPresent})
+		return
+	}
+
+	pluginURL := p.GetPluginURL()
```
Comment (Kshitij-Katiyar) [inline server/plugin/api.go:340]: no need for this variable this p.GetPluginURL() direct

File: server/plugin/client.go [hunk 9]
```diff
@@ -99,42 +100,76 @@ func (c *client) Link(body *serializers.LinkRequestPayload, mattermostUserID str
...
+
+	publisherInputs := serializers.PublisherInputs{
+		ProjectID: project.ProjectID,
+	}
+
+	consumerInputs := serializers.ConsumerInputs{
+		URL: fmt.Sprintf("%s%s?channelID=%s", strings.TrimRight(pluginURL, "/"), constants.PathNotificationSubscription, channelID),
+	}
+
+	StatusData := map[string]string{
```
Comment (Kshitij-Katiyar) [inline server/plugin/client.go:122]:
```suggestion
	statusData := map[string]string{
```

File: server/store/subscriptions.go [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+// TODO: remove later if not needed.
+func (subscriptionList *SubscriptionList) DeleteSubscriptionByKey(userID, subscriptionKey string) {
+	for key := range subscriptionList.ByMattermostUserID[userID] {
+		if key == subscriptionKey {
+			delete(subscriptionList.ByMattermostUserID[userID], key)
+		}
+	}
+}
+
+func SubscriptionListFromJSON(bytes []byte) (*SubscriptionList, error) {
```
Comment (Kshitij-Katiyar) [inline server/store/subscriptions.go:122]: move this to serializers if possible

Comment (ayusht2810) [reply server/store/subscriptions.go:122]: did the same while linking a project

File: server/plugin/api.go [hunk 11]
```diff
@@ -284,6 +285,81 @@ func (p *Plugin) handleCreateTask(w http.ResponseWriter, r *http.Request) {
...
+	}
+
+	p.Store.StoreSubscription(&serializers.SubscriptionDetails{
+		MattermostUserID: mattermostUserID,
+		ProjectName:      body.Project,
+		OrganizationName: body.Organization,
+		EventType:        body.EventType,
+		ChannelID:        channel.Id,
+	})
+	response, err := json.Marshal(subscription)
```
Comment (raghavaggarwal2308) [inline server/plugin/api.go:351]:
```suggestion

	response, err := json.Marshal(subscription)
```

File: server/plugin/api.go [hunk 12]
```diff
@@ -293,6 +369,11 @@ func (p *Plugin) testAPI(w http.ResponseWriter, r *http.Request) {
 		w.Write([]byte(err.Error()))
 		return
 	}
+	// TODO: for testing purposes, remove later
+	mattermostUserID := r.Header.Get(constants.HeaderMattermostUserIDAPI)
+	u, _ := p.Store.LoadUser(mattermostUserID)
+	t, _ := p.ParseAuthToken(u.AccessToken)
```
Comment (raghavaggarwal2308) [inline server/plugin/api.go:375]: Why not hanling errors?

File: server/plugin/api.go [hunk 13]
```diff
@@ -293,6 +369,11 @@ func (p *Plugin) testAPI(w http.ResponseWriter, r *http.Request) {
...
+	fmt.Println("\n\n\n", t)
```
Comment (raghavaggarwal2308) [inline server/plugin/api.go:376]: remove this

**Review by raghavaggarwal2308 — CHANGES_REQUESTED** [review]

File: server/plugin/api.go (hunk 12 above)
Comment (ayusht2810) [reply server/plugin/api.go]: removed this part, as it was for testing only

File: server/plugin/api.go [hunk 14]
```diff
@@ -369,11 +370,7 @@ func (p *Plugin) testAPI(w http.ResponseWriter, r *http.Request) {
 		w.Write([]byte(err.Error()))
 		return
 	}
-	// TODO: for testing purposes, remove later
-	mattermostUserID := r.Header.Get(constants.HeaderMattermostUserIDAPI)
-	u, _ := p.Store.LoadUser(mattermostUserID)
-	t, _ := p.ParseAuthToken(u.AccessToken)
-	fmt.Println("\n\n\n", t)
+
 	res, _ := json.Marshal(response)
```
Comment (raghavaggarwal2308) [inline server/plugin/api.go:374]: Why not checking error?

**Review by raghavaggarwal2308 — CHANGES_REQUESTED** [review]

File: server/plugin/api.go (hunk 14 above)
Comment (ayusht2810) [reply server/plugin/api.go:374]: removed this test API from the next PR so you can ignore it for now

**Review by raghavaggarwal2308 — APPROVED** [review]

**Comment by ayusht2810** [comment]
Created new PR for these changes #32

**Pull request closed without merging**